Update plugintemplate to CMake 3.0 build system and modern code requirements

PatTheMav opened the pull request:
### Description
This PR represents an overhaul of the entire plugin to stay compatible with changes also being introduced at the main repository, including:

* Update of CMake scripts to the new framework introduced in the `obs-studio` repository
* Update of all build flags to include stricter code requirements and more enabled warnings
* Update of all GitHub Actions workflows to use a more modular design and support schedule as well as dispatch workflow runs
* Reduce complexity of scripts by moving build dependency resolution into CMake scripts
* Remove all functionality to build `obs-studio` on Linux and rely on `libobs-dev` package instead
* Remove reliance on generated header and macro definitions as well as macro functions
* Uses optimised build configurations:
    * `Debug` for pull requests and dispatches
    * `RelWithDebInfo` for pushes on the master branch and scheduled builds
    * `Release` for tag pushes
* Releases now also contain a "source tarball"
* Added `aarch64` as available Linux build configuration (CMake toolchain scripts for cross compilation from/to `x86_64` from/to `aarch64` are included)

### Motivation and Context
Update scripts and source code to updated standards on the main repository, which also allows to share more code between both.

Closes https://github.com/obsproject/obs-plugintemplate/pull/41
Resolves https://github.com/obsproject/obs-plugintemplate/pull/45
Resolves https://github.com/obsproject/obs-plugintemplate/pull/47
Fixes https://github.com/obsproject/obs-plugintemplate/pull/55
Fixes https://github.com/obsproject/obs-plugintemplate/pull/63
Closes https://github.com/obsproject/obs-plugintemplate/pull/66
Resolves https://github.com/obsproject/obs-plugintemplate/pull/67
Fixes https://github.com/obsproject/obs-plugintemplate/pull/68


### How Has This Been Tested?
Tested pull requests, dispatches, pushes, and releases on separate fork, including macOS code signing and notarization.

### Types of changes
- New feature (non-breaking change which adds functionality)
- Code cleanup (non-breaking change which makes code smaller or more readable)
- Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

RytoEX commented on .github/actions/run-clang-format/action.yaml:
```diff
@@ -0,0 +1,48 @@
+      shell: bash
+      run: |
+        : Check Runner Operating System 🏃‍♂️
+        echo "::notice::run-clang-format action requires a macOS- or Linux-based runner."; exit 2
```
```suggestion
        echo "::notice::run-clang-format action requires a macOS-based or Linux-based runner."; exit 2
```

RytoEX commented on .github/actions/run-clang-format/action.yaml:
```diff
@@ -0,0 +1,48 @@
+description: Runs clang-format and checks for any changes introduced by it
+inputs:
+  failCondition:
+    description: Should checks result in error response
```
I'm having trouble understanding how this should be read or interpreted.

PatTheMav replied:
In essence, you can choose for the check not to result in a failed workflow run.

RytoEX commented on .github/actions/run-cmake-format/action.yaml:
```diff
@@ -0,0 +1,48 @@
+description: Runs cmake-format and checks for any changes introduced by it
+inputs:
+  failCondition:
+    description: Should checks result in error response
```
I'm having trouble understanding how this should be read or interpreted.

RytoEX commented on .github/actions/run-cmake-format/action.yaml:
```diff
@@ -0,0 +1,48 @@
+      shell: bash
+      run: |
+        : Check Runner Operating System 🏃‍♂️
+        echo "::notice::run-cmake-format action requires a macOS- or Linux-based runner."; exit 2
```
```suggestion
        echo "::notice::run-cmake-format action requires a macOS-based or Linux-based runner."; exit 2
```

RytoEX commented on .github/actions/setup-macos-codesigning/action.yaml:
```diff
@@ -0,0 +1,127 @@
+name: Setup macOS codesigning
```
```suggestion
name: Set up macOS codesigning
```

RytoEX commented on .github/workflows/pr-pull.yaml:
```diff
@@ -0,0 +1,20 @@
+name: Pull Request
+run-name: ${{ github.head_ref }} pull request run 🚀
```
Other run-name values appear to be Title Case. Where does this appear?

PatTheMav replied:
That appears as the title of the workflow run in the GitHub action tab.

RytoEX reviewed (changes requested):
A few quick notes.

PatTheMav commented:
Small update: The build system itself has support for `clang` on Windows and Linux, but it's not enabled by default:

* On Linux the `aarch64-linux-clang.cmake` and `x86_64-linux-clang.cmake` toolchain files need to be used to switch
* On Windows `-T clangcl` needs to be added to the `cmake` configuration command line to enable it 

I also removed the mandatory flags to treat warnings as errors, as CMake has introduced the `CMAKE_COMPILE_WARNING_AS_ERROR` boolean variable in CMake 3.24 that allows switching this behaviour on and off - so CI builds have it enabled, but local development does not have, which should be a nice middle ground.

tytan652 commented on cmake/common/osconfig.cmake:
```diff
@@ -0,0 +1,17 @@
+
+# Set minimum CMake version specific to host operating system, add OS-specific module directory to default search paths,
+# and set helper variables for OS detection in other CMake list files.
+if(CMAKE_HOST_SYSTEM_NAME STREQUAL "Windows")
+  list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/windows")
+  set(OS_WINDOWS TRUE)
+elseif(CMAKE_HOST_SYSTEM_NAME STREQUAL "Darwin")
+  list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/macos")
+  set(OS_MACOS TRUE)
+elseif(CMAKE_HOST_SYSTEM_NAME MATCHES "Linux|FreeBSD|OpenBSD")
+  list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/linux")
+  string(TOUPPER "${CMAKE_HOST_SYSTEM_NAME}" _SYSTEM_NAME_U)
+  set(OS_${_SYSTEM_NAME_U} TRUE)
+endif()
```
Don't enable/add cross-compile if you rely on `CMAKE_HOST_` variable.

PatTheMav replied:
That is actually the desired behaviour here - the decisions need to be made based on the _host_ architecture, not the _target_ architecture. We do not support cross compiling for other platforms, only for other architectures on the same platform.

Cross compilation with CMake should _never_ change the host variables (hence why they are called `CMAKE_HOST_<SOMETHING>`.

tytan652 replied:
I'm just saying that `CMAKE_SYSTEM_NAME` must be used there. It would makes no sense to use the host system name while cross-compiling, the toolchain set `CMAKE_SYSTEM_NAME` for a reason.

PatTheMav replied:
It can set it all it wants, the bootstrapper _requires_ the host system. It is a conscious choice and not a mistake. Which is also why the bootstrapper runs _before_ the project call - at this point in execution there is no notion of a project or a compilation target, just the host system the CMake script runs on.

tytan652 replied:
So you voluntarily prevent people to be able to test-compile for other platform (not create a bundle just test if it compiles) through cross-compilation.

RytoEX commented on .github/workflows/build-project.yaml:
```diff
@@ -0,0 +1,219 @@
+        with:
+          submodules: recursive
+          fetch-depth: 0
+      - name: Setup Environment 🔧
```
Nit: "Set up" or "Set Up"

"Setup" is a noun.

RytoEX commented on .github/workflows/build-project.yaml:
```diff
@@ -0,0 +1,219 @@
+        with:
+          submodules: recursive
+          fetch-depth: 0
+      - name: Setup Environment 🔧
```
Nit: "Set up" or "Set Up"

RytoEX commented on .github/workflows/build-project.yaml:
```diff
@@ -0,0 +1,219 @@
+      - name: Setup Environment 🔧
+        id: setup
+        run: |
+          : Setup Environment 🔧
```
Nit: "Set up" or "Set Up"

RytoEX commented on .github/workflows/build-project.yaml:
```diff
@@ -0,0 +1,219 @@
+        with:
+          submodules: recursive
+          fetch-depth: 0
+      - name: Setup Environment 🔧
```
Nit: "Set up" or "Set Up"

RytoEX commented on .github/workflows/build-project.yaml:
```diff
@@ -0,0 +1,219 @@
+      - name: Setup Environment 🔧
+        id: setup
+        run: |
+          # Setup Environment 🔧
```
Nit: "Set up" or "Set Up"

RytoEX reviewed (changes requested):
Another review round before I forget to submit it. I'm about two-thirds through the changed files.

RytoEX commented on src/plugin-support.h:
```diff
@@ -16,12 +16,15 @@ You should have received a copy of the GNU General Public License along
+#include <string.h>
 
-#define blog(level, msg, ...) blog(level, "[" PLUGIN_NAME "] " msg, ##__VA_ARGS__)
+extern const char *PLUGIN_VERSION;
+extern const char *PLUGIN_NAME;
```
Nit: prefer NAME then VERSION.

RytoEX commented on src/plugin-support.c.in:
```diff
@@ -0,0 +1,39 @@
+
+#include <plugin-support.h>
+
+const char *PLUGIN_VERSION = "@CMAKE_PROJECT_VERSION@";
+const char *PLUGIN_NAME = "@CMAKE_PROJECT_NAME@";
```
Nit: prefer NAME then VERSION.

RytoEX commented on src/plugin-support.c.in:
```diff
@@ -0,0 +1,39 @@
+
+#include <plugin-support.h>
+
+const char *PLUGIN_VERSION = "@CMAKE_PROJECT_VERSION@";
+const char *PLUGIN_NAME = "@CMAKE_PROJECT_NAME@";
```
I think in obs-studio we wanted to make these `static const char *`. Is there a practical difference here?

RytoEX replied:
Thanks. I always forget the particulars.

RytoEX commented on cmake/windows/compilerconfig.cmake:
```diff
@@ -0,0 +1,35 @@
+endif()
+
+if(CMAKE_VS_WINDOWS_TARGET_PLATFORM_VERSION VERSION_LESS 10.0.20348)
+  message(FATAL_ERROR "OBS required Windows 10 SDK version 10.0.20348.0 or more recent.\n"
```
Nit: "required" or "requires"?

RytoEX commented on cmake/macos/xcode.cmake:
```diff
@@ -0,0 +1,173 @@
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN__DUPLICATE_METHOD_MATCH YES)
+
+# Set default warnings for C and C++
+set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_ABOUT_MISSING_FIELD_INITIALIZERS NO)
+set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_ABOUT_MISSING_NEWLINE YES)
+set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_ABOUT_RETURN_TYPE YES_ERROR)
+set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_CHECK_SWITCH_STATEMENTS YES)
+set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_FOUR_CHARACTER_CONSTANTS YES)
+set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_SIGN_COMPARE YES)
+set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_SHADOW NO)
+set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_UNUSED_FUNCTION NO)
+set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_UNUSED_PARAMETER YES)
+set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_UNUSED_VALUE YES)
+set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_UNUSED_VARIABLE YES)
+set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_TYPECHECK_CALLS_TO_PRINTF YES)
+set(CMAKE_XCODE_ATTRIBUTE_GCC_NO_COMMON_BLOCKS YES)
+set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_64_TO_32_BIT_CONVERSION YES)
+set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_UNDECLARED_SELECTOR YES)
+set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_UNINITIALIZED_AUTOS YES)
```
Would it make sense to alphabetize or group these for ease of skimming?

PatTheMav replied:
The order is the result of adding additional flags later - in the Xcode project file they are alphabetised, so it's easier to just reorder them here accordingly.

RytoEX commented on cmake/macos/xcode.cmake:
```diff
@@ -0,0 +1,173 @@
+set(CMAKE_XCODE_ATTRIBUTE_ENABLE_STRICT_OBJC_MSGSEND YES)
+
+# Set default warnings for ObjC and C++
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_BLOCK_CAPTURE_AUTORELEASING YES_ERROR)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_SUSPICIOUS_IMPLICIT_CONVERSION NO)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_DEPRECATED_OBJC_IMPLEMENTATIONS YES)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_OBJC_REPEATED_USE_OF_WEAK YES)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_RANGE_LOOP_ANALYSIS YES)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_EMPTY_BODY YES)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_BOOL_CONVERSION YES)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_COMMA YES)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_CONSTANT_CONVERSION YES)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_ENUM_CONVERSION YES)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_INFINITE_RECURSION YES)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_INT_CONVERSION YES)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_NON_LITERAL_NULL_CONVERSION YES)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_OBJC_IMPLICIT_RETAIN_SELF YES)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_OBJC_LITERAL_CONVERSION YES)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_QUOTED_INCLUDE_IN_FRAMEWORK_HEADER YES)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_STRICT_PROTOTYPES NO)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_SUSPICIOUS_MOVE YES)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_UNREACHABLE_CODE YES)
+set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN__DUPLICATE_METHOD_MATCH YES)
```
Would it make sense to alphabetize or group these for ease of skimming?

PatTheMav replied:
See above.

RytoEX commented on cmake/macos/xcode.cmake:
```diff
@@ -0,0 +1,173 @@
+# Disable strict aliasing
+set(CMAKE_XCODE_ATTRIBUTE_GCC_STRICT_ALIASING NO)
+
+# Set C++ language default to c17
+#
+# * CMake explicitly sets the version via compiler flag when transitive dependencies require specific compiler feature
+#   set, resulting in the flag being added twice. Report to KitWare as a feature request for Xcode generator
+#
+# set(CMAKE_XCODE_ATTRIBUTE_GCC_C_LANGUAGE_STANDARD c17)
+#
+# Set C++ language default to c++17
+#
+# * See above. Report to KitWare as a feature request for Xcode generator
+#
+# set(CMAKE_XCODE_ATTRIBUTE_CLANG_CXX_LANGUAGE_STANDARD c++17)
```
Does this still need reported?

PatTheMav replied:
It is reported and a known issue, there has been debate about being able to declare the transitive requirements for GNU extensions, but no further developments yet.

For the time being, whenever a CMake package requires a C or C++ language standard and the current compiler supports GNU extensions, the transitive requirement will make the consumer use GNU extensions as well.

RytoEX replied:
If there's an open Kitware GitLab link, it would be good to provide the link here so that the comment doesn't indicate that we need to take that action.

RytoEX commented on cmake/linux/compilerconfig.cmake:
```diff
@@ -0,0 +1,91 @@
+  set(CMAKE_COLOR_DIAGNOSTICS ON)
+endif()
+
+if(CMAKE_CXX_COMPILER_ID STREQUAL GNU)
+  # Disable false-positive warning in GCC 12.1.0
+  if(CMAKE_CXX_COMPILER_VERSION VERSION_EQUAL 12.1.0)
+    add_compile_options(-Wno-error=maybe-uninitialized)
+  endif()
+  # Add warning for infinite recursion (added in GCC 12)
+  if(CMAKE_C_COMPILER_VERSION VERSION_GREATER_EQUAL 12.0.0)
+    add_compile_options(-Winfinite-recursion)
+  endif()
+endif()
```
Should this adapt the changes in https://github.com/obsproject/obs-studio/pull/8853?

PatTheMav replied:
I adapted them partially - disabling the type-limits warning should be fine and can be added if a plugin actually runs afoul of it. Usually the plugintemplate should not need to disable as many warnings as the main repo and it should fall upon plugin authors to extend this as they need.

By that token it might be more correct to even remove some of the silencing we do here, but I'm fine with it for now.

RytoEX commented on cmake/common/ccache.cmake:
```diff
@@ -0,0 +1,22 @@
+endif()
+
+if(CCACHE_PROGRAM)
+  message(DEBUG "CCache found as ${CCACHE_PROGRAM}...")
```
In strings, use "Ccache" if at the start of a sentence, "ccache" otherwise.

RytoEX reviewed (changes requested):
Another quick pass while I had it in mind.

PatTheMav commented on .github/scripts/.build.zsh:
```diff
@@ -50,44 +58,43 @@ build() {
   typeset -g -a skips=()
-  local -i _verbosity=1
-  local -r _version='1.0.0'
+  local -i verbosity=1
```
I removed underscores for mutable variables - only immutable ones (if possible declared as readonly via `local -r`) are supposed to carry the underscore.

PatTheMav commented on .github/scripts/.build.zsh:
```diff
@@ -98,7 +105,7 @@ Usage: %B${functrace[1]%:*}%b <option> [<options>]
   local -a args
   while (( # )) {
     case ${1} {
-      -t|--target|-c|--config|--generator)
+      -t|--target|--generator|-c|--config)
```
I originally removed the generator variable entirely but re-added it later.

PatTheMav commented on .github/scripts/.build.zsh:
```diff
@@ -122,32 +129,34 @@ Usage: %B${functrace[1]%:*}%b <option> [<options>]
         shift 2
         ;;
       -c|--config)
-        if (( ! ${_valid_configs[(Ie)${2}]} )) {
+        if (( !${_valid_configs[(Ie)${2}]} )) {
```
Nope, that's a mistake.

PatTheMav commented on .github/scripts/Build-Windows.ps1:
```diff
@@ -46,56 +44,58 @@ function Build {
-        Log-Debug "Attempting to configure OBS with CMake arguments: $($CmakeArgs | Out-String)"
-        Log-Information "Configuring ${ProductName}..."
-        Invoke-External cmake -S . -B build_${script:Target} @CmakeArgs
+        $CmakeBuildArgs = @(
```
Correct, fixed already.

PatTheMav commented on .github/scripts/Build-Windows.ps1:
```diff
@@ -46,56 +44,58 @@ function Build {
 
-        $CmakeArgs = @(
-            '--config', "${Configuration}"
+        $CmakeInstallArgs = @(
```
See above.

PatTheMav commented on .github/scripts/utils.zsh/setup_linux:
```diff
@@ -13,50 +13,49 @@ if (( ! ${+target} )) {
 
 if (( ! (${skips[(Ie)all]} + ${skips[(Ie)deps]}) )) {
-  log_info 'Installing obs build dependencies...'
+  log_group 'Installing obs build dependencies...'
```
Should either be `OBS` or `obs-studio`.

gxalpha commented:
Hi, just tried this on macOS, and noticed two things:
- libobs' `obs.hpp` C++ helper header is not available in the libobs framework used here.
- With `ENABLE_QT=ON` I get a bunch of warnings-turned-into-errors inside of Qt (`Possible misuse of comma operator here`, `Double-quoted include "qapplication.h" in framework header, expected angle-bracketed instead`, etc). It might make sense to disable warnings-as-errors just in Qt if that's somehow possible. I worked around this for now by setting `-DCMAKE_COMPILE_WARNING_AS_ERROR=OFF` 

Seeing as I couldn't use the `AutoRelease` types because of the missing header I can't say much more than "the plugin built and got loaded and would probably work".

gxalpha commented on .github/workflows/scheduled.yaml:
In addition to my comments from yesterday, I wonder if this nightly(?) scheduled job is a good idea to have enabled by default.

I at least wasn't aware of it at all until I received an email about it failing this morning, and I suspect many plugin authors won't be aware of it either. Especially for plugins which often times don't get updated every day (the majority of them receives new commits rather rarely, as well-scoped plugins tend to "just work" after they're made), this seems a bit wasteful, and becomes a problem when they're hosted on private repos (e.g. while still under initial development, when they're created for specific people and not meant to be generally available, etc.) and just eat away GitHub Actions minutes.

RytoEX commented on cmake/common/ccache.cmake:
```diff
@@ -0,0 +1,22 @@
+  option(ENABLE_CCACHE "Enable compiler acceleration with ccache" ON)
+
+  if(ENABLE_CCACHE)
+    set(CMAKE_CXX_COMPILER_LAUNCHER "${CCACHE_PROGRAM}")
+    set(CMAKE_C_COMPILER_LAUNCHER "${CCACHE_PROGRAM}")
```
Nit: C then CXX

RytoEX commented on cmake/common/buildspec_common.cmake:
```diff
@@ -0,0 +1,215 @@
+        # cmake-format: on
+        continue()
+      else()
+        message(AUTHOR_WARNING "Future ${label} version detected in ${path}: \n"
```
"Newer" instead of "Future"?

RytoEX commented on cmake/common/buildspec_common.cmake:
```diff
@@ -0,0 +1,215 @@
+        # cmake-format: on
+        break()
+      elseif(_check_version VERSION_LESS version)
+        message(AUTHOR_WARNING "Outdated ${label} version detected in ${path}: \n"
```
"Older" instead of "Outdated"? I feel less strongly about this than the "Future" line below.

RytoEX commented on buildspec.json:
```diff
@@ -1,88 +1,67 @@
     "version": "1.0.0",
     "author": "Your Name Here",
-    "email": "me@mymailhost.com"
+    "website": "https://www.your-website.tld",
+    "email": "me@mymailhost.com",
```
We should probably use `example.com` here.

PatTheMav replied:
I specifically used this because of the recent `example.com` possibly being a malicious site discussion.

gxalpha replied:
There must have been some confusion here. mywebsite.com was the possibly malicious one, example.com is IANA reserved for the purpose of being an example domain (like here).

RytoEX replied:
> There must have been some confusion here. mywebsite.com was the possibly malicious one, example.com is IANA reserved for the purpose of being an example domain (like here).

What @gxalpha said. `example.com` is specifically reserved and held by the IANA for the purpose of using as an example domain.

RytoEX commented on buildspec.json:
```diff
@@ -1,88 +1,67 @@
+            "label": "OBS sources",
             "hashes": {
-                "macos-x86_64": "7637e52305e6fc53014b5aabd583f1a4490b1d97450420e977cae9a336a29525",
-                "macos-arm64": "755e0fa69b17a3ae444e1befa9d91d77e3cafe628fbd1c6333686091826595cd",
-                "macos-universal": "de057e73e6fe0825664c258ca2dd6798c41ae580bf4d896e1647676a4941934a",
-                "windows-x64": "2192d8ce780c4281b807cd457994963669e5202659ecd92f19b54c3e7d0c1915",
-                "windows-x86": "9f8582ab5891b000869d6484ea591add9fbac9f1c91b56c7b85fdfd56a261c1b"
+                "macos": "61ff20291c3f1e24ad7dfbc8c7078258a0141b15e749d7020143fa572da17d63",
+                "windows-x64": "93e3055358198c54c1d87e3c6a1600fed2b349d41335faf016e15473da0c8ce9"
```
Can probably change this to 29.1.0 or 29.1.1. I do wonder if it would be "less disruptive" to simply change the versions and hashes for buildspec in a separate small PR and keep changes in this PR to ones necessary for the broader updates.

PatTheMav replied:
Yeah I updated this to 29.1.1.

RytoEX commented on build-aux/.functions/log_group:
```diff
@@ -0,0 +1,16 @@
+    print "::endgroup::"
+    typeset -g _log_group=0
+  }
+  if (( # )) {
```
Is this a zsh-ism?

PatTheMav replied:
What specifically?

PatTheMav replied:
Yeah that's ZSH - `${#}` gives you the number of positional arguments to a function and arithmetic checks in ZSH do not need the variable notation (`if (( VAR ))` is equal in function to `if (( ${VAR} ))`).

RytoEX commented on CMakePresets.json:
```diff
@@ -0,0 +1,177 @@
+      "configurePreset": "windows-x64",
+      "displayName": "Windows x64",
+      "description": "Windows build for x64",
+      "configuration": "Release"
```
Do we actually encourage plugin authors to release plugins in Release instead of RelWithDebInfo?

PatTheMav replied:
Switched back to RelWithDebInfo for now - macOS/Xcode enforces Release for archiving, CPack's `DEB` generator automatically strips binaries, so those platforms behave differently than Windows.

RytoEX commented on .github/scripts/utils.zsh/read_codesign_team:
```diff
@@ -0,0 +1,7 @@
+
+if (( ! ${+CODESIGN_TEAM} )) {
+  typeset -g CODESIGN_TEAM
+  log_info 'Setting up developer team id for codesigning...'
```
Nit: s/id/ID/

Perhaps "Developer Team ID"?

RytoEX commented on .github/scripts/Package-Windows.ps1:
```diff
@@ -14,14 +15,21 @@ if ( $DebugPreference -eq 'Continue' ) {
 }
 
+if ( ! ( [System.Environment]::Is64BitOperatingSystem ) ) {
+    throw "Packaging script requires a 64bit system to build and run."
```
Nit: 64-bit

RytoEX commented on .github/actions/build-plugin/action.yaml:
```diff
@@ -0,0 +1,95 @@
+        ${{ inputs.workingDirectory }}/.github/scripts/build-linux ${build_args}
+
+        local -a ccache_data=("${(fA)$(ccache -s -vv)}")
+        print '### Ubuntu CCache Stats (${{ inputs.target }})' >> $GITHUB_STEP_SUMMARY
```
Nit: Ccache

(Title Case)

RytoEX commented on .github/actions/build-plugin/action.yaml:
```diff
@@ -0,0 +1,95 @@
+        ${{ inputs.workingDirectory }}/.github/scripts/build-macos ${build_args}
+
+        local -a ccache_data=("${(fA)$(ccache -s -vv)}")
+        print '### macOS CCache Stats (${{ inputs.target }})' >> $GITHUB_STEP_SUMMARY
```
Nit: Ccache

(Title Case)

RytoEX reviewed (changes requested):
Another review.

RytoEX commented on .github/workflows/build-project.yaml:
```diff
@@ -0,0 +1,258 @@
+          case "${GITHUB_EVENT_NAME}" in
+            pull_request)
+              config_data=('codesign:false' 'notarize:false' 'package:false' 'config:Debug')
+              if curl -H "Authorization: Bearer ${{ secrets.GITHUB_TOKEN }}" -s "${{ github.event.pull_request.url }}" \
+                | jq -e -r '.labels[] | select(.name == "Seeking Testers")'; then
```
Bare `if`?

PatTheMav replied:
That is correct because just the return value of the command is checked - you only use `test` (or `[[ ]]`) if you want to check values. In this case we rely on the return code of `jq` so `if <command>; then` is the usual pattern.

RytoEX commented on .github/workflows/build-project.yaml:
```diff
@@ -0,0 +1,258 @@
+
+  macos-build:
+    name: Build for macOS 🍏
+    runs-on: macos-13
```
Are these runners still in beta? Do we feel confident using them here?

RytoEX replied:
Is there any possible issue for plugins being built on newer runners/Xcode with obs-studio being built on macos-12 runners (and associated Xcode version)?

PatTheMav replied:
None that I'm aware of - on the contrary, building on macOS 13/Xcode 14.3 should lead to _less_ issues as its the most recent release, which one should always be on.

RytoEX commented on .github/workflows/build-project.yaml:
```diff
@@ -0,0 +1,258 @@
+          codesignUser: ${{ secrets.MACOS_NOTARIZATION_USERNAME }}
+          codesignPass: ${{ secrets.MACOS_NOTARIZATION_PASSWORD }}
+
+      - name: Upload artifacts 📡
```
```suggestion
      - name: Upload Artifacts 📡
```

PatTheMav replied:
Probably requires a title-case pass over all the names.

RytoEX commented on .github/workflows/build-project.yaml:
```diff
@@ -0,0 +1,258 @@
+          restore-keys: |
+            ${{ runner.os }}-ccache-x86_64-
+
+      - name: Set up Homebrew 🍺
```
```suggestion
      - name: Set Up Homebrew 🍺
```

RytoEX commented on .github/workflows/build-project.yaml:
```diff
@@ -0,0 +1,258 @@
+          target: x86_64
+          config: ${{ needs.check-event.outputs.config }}
+
+      - name: Upload source tarball 🗜️
```
```suggestion
      - name: Upload Source Tarball 🗜️
```

RytoEX commented on .github/workflows/build-project.yaml:
```diff
@@ -0,0 +1,258 @@
+          name: ${{ steps.setup.outputs.pluginName }}-${{ steps.setup.outputs.pluginVersion }}-sources-${{ needs.check-event.outputs.commitHash }}
+          path: ${{ github.workspace }}/release/${{ steps.setup.outputs.pluginName }}-${{ steps.setup.outputs.pluginVersion }}-source.*
+
+      - name: Upload artifacts 📡
```
```suggestion
      - name: Upload Artifacts 📡
```

RytoEX commented on .github/workflows/build-project.yaml:
```diff
@@ -0,0 +1,258 @@
+          name: ${{ steps.setup.outputs.pluginName }}-${{ steps.setup.outputs.pluginVersion }}-ubuntu-22.04-x86_64-${{ needs.check-event.outputs.commitHash }}
+          path: ${{ github.workspace }}/release/${{ steps.setup.outputs.pluginName }}-${{ steps.setup.outputs.pluginVersion }}-x86_64*.*
+
+      - name: Upload debug symbol artifacts 🪲
```
```suggestion
      - name: Upload Debug Symbol Artifacts 🪲
```

RytoEX commented on .github/workflows/build-project.yaml:
```diff
@@ -0,0 +1,258 @@
+          config: ${{ needs.check-event.outputs.config }}
+          package: ${{ fromJSON(needs.check-event.outputs.package) }}
+
+      - name: Upload artifacts 📡
```
```suggestion
      - name: Upload Artifacts 📡
```

RytoEX commented on .github/workflows/push.yaml:
```diff
@@ -0,0 +1,114 @@
+            *) echo 'validTag=false' >> $GITHUB_OUTPUT ;;
+          esac
+
+      - name: Download build artifacts 📥
```
```suggestion
      - name: Download Build Artifacts 📥
```

RytoEX commented on build-aux/.run-format.zsh:
```diff
@@ -0,0 +1,188 @@
+  typeset -g host_os=${${(L)$(uname -s)}//darwin/macos}
+  local -i fail_on_error=0
+  local -i check_only=0
+  local -i _verbosity=1
```
Should this still be prefixed with an underscore?

RytoEX reviewed (changes requested):
Finally finished checking all files. @gxalpha if you're running this on a plugin repo, I'd appreciate additional feedback or approval.

I did see places where strings with variables are not quoted and did not flag them. Does zsh handle those cases without requiring quotes? Or are they cases where we have a reasonable degree of certainty that the strings won't contain whitespace (e.g., GitHub Actions runner paths)?

PatTheMav commented on cmake/windows/helpers.cmake:
```diff
@@ -0,0 +1,130 @@
+
+  while(_STPO_PROPERTIES)
+    list(POP_FRONT _STPO_PROPERTIES key value)
+    set_property(TARGET ${target} PROPERTY ${key} "${value}")
```
Yeah keys can never contain spaces, if they do it should fail.

PatTheMav commented on cmake/common/compiler_common.cmake:
```diff
@@ -0,0 +1,89 @@
+    -Wimplicit-retain-self
+    -Winvalid-offsetof
+    -Wmove
+    -Wno-c++11-extensions
```
That's the default flag used by Xcode, we're just making it used by all `clang`-based compilations.

RytoEX replied:
I have to wonder at some point if it would be better to just use `-pedantic`, but that's perhaps a question for another day. This is fine as-is, if not a bit oddly specific.

PatTheMav commented on .github/scripts/Build-Windows.ps1:
```diff
@@ -19,6 +16,10 @@ if ( $DebugPreference -eq 'Continue' ) {
 }
 
+if ( ! ( [System.Environment]::Is64BitOperatingSystem ) ) {
+    throw "obs-studio requires a 64bit system to build and run."
```
In a way both, but not directly - it will build the plugin, but the plugin's CMake script will build `libobs` automatically (at least on Windows and macOS).

gxalpha commented:
> @gxalpha if you're running this on a plugin repo, I'd appreciate additional feedback or approval

Sure, I’ll give it a further look when I have time (might not be until Thursday).

> Needs to be fixed in obs-studio.

Very unfortunate, I suspect as-is this would block many developers whose code base already uses the types from switching to the new template. How difficult would fixing it be (or is it by any chance part of one of your PRs already)?

> Cannot reproduce - how do you end up in this state?

I think I just used the macos cmake preset, but will double-check when I have time (again, perhaps not until Thursday). It probably requires including (and maybe using) the headers in the source file.

PatTheMav commented:
> > @gxalpha if you're running this on a plugin repo, I'd appreciate additional feedback or approval
> 
> Sure, I’ll give it a further look when I have time (might not be until Thursday).
> 
> > Needs to be fixed in obs-studio.
> 
> Very unfortunate, I suspect as-is this would block many developers whose code base already uses the types from switching to the new template. How difficult would fixing it be (or is it by any chance part of one of your PRs already)?

Easy to fix - it's a one-liner PR.

gxalpha commented:
> Easy to fix - it's a one-liner PR.

Would be worth including in a potential 29.1.2 then imo

PatTheMav commented:
> > Easy to fix - it's a one-liner PR.
> 
> Would be worth including in a potential 29.1.2 then imo

https://github.com/obsproject/obs-studio/pull/8917

gxalpha commented on buildspec.json:
```diff
@@ -1,88 +1,67 @@
     "author": "Your Name Here",
-    "email": "me@mymailhost.com"
+    "website": "https://example.com",
+    "email": "me@mymailhost.com",
```
We'd probably also want to replace `mymailhost.com` with `example.com`

RytoEX commented on cmake/common/compiler_common.cmake:
```diff
@@ -0,0 +1,89 @@
+  set(CMAKE_C_STANDARD 11)
+endif()
+set(CMAKE_C_STANDARD_REQUIRED TRUE)
+set(CMAKE_C_EXTENSIONS FALSE)
```
Didn't this cause issues on Linux/POSIX?

PatTheMav replied:
I'm not aware of any. I've compiled OBS itself with that setting without issue, albeit only on Ubuntu.

RytoEX replied:
See:
* https://github.com/obsproject/obs-studio/blob/29.1.1/cmake/Modules/CompilerConfig.cmake#L17-L20
* https://github.com/obsproject/obs-studio/pull/8684

PatTheMav replied:
Thanks, that allowed me to investigate the root causes for that:

It seems that on Linux `glibc` more or less relies on GNU extensions for a lot of core functionality. On obs-studio we merged code in `darray.h` that _requires_ them even, so disabling extensions on Linux will not be possible.

Thus I limited that setting to macOS and Windows.

RytoEX replied:
Ah, I see now.
* the darray test code in `darray.h` requires C extensions and is ifdef'd to only be included on Linux
* thus it fails on Linux, because we tried to globally disable C extensions and the code requires them

That is _unfortunate_.

RytoEX replied:
On closer look, it seems like that code should only be hit `if(CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX)` and the builder/user/dev also left `ENABLE_DARRAY_TYPE_TEST` set to `ON` (which admittedly is default here).

Is this file used for settings to build OBS itself or just the plugin? If it's _just_ the plugin, then we could probably leave C extensions disabled across the board to discourage extension usage. Might we be able to check the value of `ENABLE_DARRAY_TYPE_TEST` to toggle it on and off? Otherwise, I suppose we can leave `set(CMAKE_C_EXTENSIONS FALSE)` limited to macOS and Windows.

PatTheMav replied:
The header itself (as with all the other headers in libobs) is unfortunately always imported by libobs' headers (I'd argue we expose/force to include way too much stuff, but libobs' header infrastructure has never been designed with standalone usage in mind).

As such the code and lots of other stuff is automatically added to a plugin's codebase and requires the same flags.

PatTheMav replied:
FWIW it seems `threading.h` in the utils directory also requires GNU extensions to be enabled on Linux, as some of the `pthread` functionality we use is only enabled in that header if `gcc` runs in a non-standards mode.

PatTheMav commented on .github/actions/build-plugin/action.yaml:
```diff
@@ -0,0 +1,98 @@
+name: 'Setup and build plugin'
+description: 'Builds the plugin for specified architecture and build config.'
+inputs:
+  target:
+    description: 'Build target for dependencies'
```
The actions _builds_ the plugin, but `target` sets the "build target" - probably needs hyphenation ("Build-target")?

RytoEX replied:
Ah, if "build target" is meant to be a noun, then a hyphen would probably help. I was reading it as a verb. Or something that adds clarity:
* The build target for dependencies
* Specify the build target for dependencies

Or some such. Will leave it to you.

PatTheMav replied:
Replaced it with "Target architecture".

PatTheMav commented on .github/actions/setup-macos-codesigning/action.yaml:
```diff
@@ -0,0 +1,154 @@
+
+          print -n "${MACOS_SIGNING_CERT}" | base64 --decode --output="${certificate_path}"
+
+          : "${MACOS_KEYCHAIN_PASSWORD:="$(LC_CTYPE=C LC_ALL=C tr -dc a-zA-Z0-9-_ < /dev/urandom | head -c 32)"}"
```
Different repo, but should be a copy-paste error.

PatTheMav commented on .github/scripts/utils.zsh/read_codesign:
```diff
@@ -2,6 +2,8 @@ autoload -Uz log_info
 if (( ! ${+CODESIGN_IDENT} )) {
   typeset -g CODESIGN_IDENT
-  log_info 'Setting up identity for application codesigning...'
+  log_info 'Setting up Apple Developer ID for application codesigning...'
```
It needs to be "a" Apple Developer ID (one can have multiple set up in the developer profile).

RytoEX commented on .github/scripts/utils.zsh/check_linux:
```diff
@@ -1,4 +1,19 @@
+log_group 'Check Linux build requirements'
+log_debug 'Checking Linux distribution name and version...'
+
+typeset -g -i UBUNTU_2210_OR_LATER=0
```
Remind me what the significance of UBUNTU_2210_OR_LATER is? Some specific feature we need to detect?

PatTheMav replied:
`libsrt` and `librist` are both available in Ubuntu 22.10, but not 22.04. As the file is a straight clone of the variant used for obs-studio in the future, it was left in to make updates/staying in sync with the main repo easier.

RytoEX replied:
Could a comment to that effect be left here? I was struggling to recall why this specific version was significant.

PatTheMav replied:
Added comment.

RytoEX commented on cmake/common/osconfig.cmake:
```diff
@@ -0,0 +1,21 @@
+# and set helper variables for OS detection in other CMake list files.
+if(CMAKE_HOST_SYSTEM_NAME STREQUAL "Windows")
+  set(CMAKE_C_EXTENSIONS FALSE)
+  set(CMAKE_CXX_EXTENSIONS FALSE)
+  list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/windows")
+  set(OS_WINDOWS TRUE)
+elseif(CMAKE_HOST_SYSTEM_NAME STREQUAL "Darwin")
+  set(CMAKE_C_EXTENSIONS FALSE)
+  set(CMAKE_CXX_EXTENSIONS FALSE)
```
As far as I know, we didn't have any issues with disabling C++ extensions on Linux. It's only C extensions that we have an issue with, as previously discussed with the `darray` tests. We should be able to globally disable C++ extensions, unless there's some other issue with that.

PatTheMav replied:
Got it - in that case I'd just add it to the Linux branch below as it makes more sense to keep that setting in a single file and not distribute it further.

RytoEX reviewed (commented):
Some additional comments.

gxalpha commented:
> Cannot reproduce - how do you end up in this state?

@PatTheMav These seem to happen when you have includes like `#include <QtWidgets/QDockWidget>` instead of `#include <QDockWidget>` (the latter doesn't throw warnings). Unfortunately this is what Qt UIC uses when generating the `ui_[...].h` headers.

gxalpha commented on buildspec.json:
```diff
@@ -1,88 +1,67 @@
+    "website": "https://example.com",
+    "email": "me@example.com",
+    "uuids": {
+        "macosPackage": "0B7A74BC-65CF-4FF1-AC34-5C743E8A48F5",
+        "macosInstaller": "52B6084A-F58D-45C3-BE37-76AD45F16072",
+        "windowsApp": "CD703FE5-1F2C-4837-BD3D-DD840D83C3E3"
```
Are these UUIDs meant to be generated by the plugin author?
It might make sense to replace them with `XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX` or something invalid and throwing a warning if it's not changed, I can envision many plugin authors not changing them otherwise (because they might not realize).

PatTheMav replied:
It would make sense to replace this with a placeholder (which needs replacing) but it also needs to be valid, otherwise CI on this repo would fail.

But if we use a valid value here, there's the risk of plugins re-using the same UUID.

RytoEX replied:
The only thing I can think of is to provide a valid UUID that is obviously "wrong" looking (all zeroes or some such), and writing some kind of check that will allow this on obsproject/obs-plugintemplate CI, but warn/error/fail on any other CI and give instructions on how to generate replacement UUIDs.

PatTheMav replied:
That wouldn't prohibit their "bad" UUIDs from also work on their own CI runs, except if we _also_ add checks if its our repo and at that point I'd question if that is all worth the effort.

Using a UUID entirely of `0` or `F` seems good enough as a possible alternative, because at some point developers just need to know what they're doing.

RytoEX replied:
> except if we _also_ add checks if its our repo and at that point I'd question if that is all worth the effort.

That is more or less what I was describing, yes.

gxalpha replied:
> Using a UUID entirely of `0` or `F` seems good enough as a possible alternative, because at some point developers just need to know what they're doing.

Apologies for only responding now (I was a bit busy this evening), but yes, this was my idea behind suggesting `XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX` (or which ever character you want to use), it’s something that much more obviously needs changing than an already set "normal" UUID.

gxalpha commented on .github/workflows/push.yaml:
```diff
@@ -0,0 +1,121 @@
+  push:
+    branches:
+      - master
+      - 'release/**'
```
The default branch name by GitHub (and maybe even git?) is `main` now so it makes sense to include that here (and in other places, e.g. the lines below and the pr-pull file) as well.

PatTheMav commented:
> > Cannot reproduce - how do you end up in this state?
> 
> @PatTheMav These seem to happen when you have includes like `#include <QtWidgets/QDockWidget>` instead of `#include <QDockWidget>` (the latter doesn't throw warnings). Unfortunately this is what Qt UIC uses when generating the `ui_[...].h` headers.

Ah yes, Qt generates lots of stuff that is actually wrong on macOS, but we also need to suppress those on obs-studio.

I added suppression for both types of warnings as there is little chance those will be fixed upstream anytime soon.

RytoEX commented:
> > > Cannot reproduce - how do you end up in this state?
> > 
> > 
> > @PatTheMav These seem to happen when you have includes like `#include <QtWidgets/QDockWidget>` instead of `#include <QDockWidget>` (the latter doesn't throw warnings). Unfortunately this is what Qt UIC uses when generating the `ui_[...].h` headers.
> 
> Ah yes, Qt generates lots of stuff that is actually wrong on macOS, but we also need to suppress those on obs-studio.
> 
> I added suppression for both types of warnings as there is little chance those will be fixed upstream anytime soon.

Is there an open QTBUG at all documenting that their UIC generation is incorrect?

RytoEX commented:
https://github.com/obsproject/obs-plugintemplate/compare/b8048701caefc069b40cd9fafbbd21e4f8233823..c5963d2f1ec5b815648017a4de4b44b63cdd8451

It's C Extensions, not CXX Extensions, that we need to disable on Linux/POSIX.

PatTheMav commented:
> > > > Cannot reproduce - how do you end up in this state?
> > > 
> > > 
> > > @PatTheMav These seem to happen when you have includes like `#include <QtWidgets/QDockWidget>` instead of `#include <QDockWidget>` (the latter doesn't throw warnings). Unfortunately this is what Qt UIC uses when generating the `ui_[...].h` headers.
> > 
> > 
> > Ah yes, Qt generates lots of stuff that is actually wrong on macOS, but we also need to suppress those on obs-studio.
> > I added suppression for both types of warnings as there is little chance those will be fixed upstream anytime soon.
> 
> Is there an open QTBUG at all documenting that their UIC generation is incorrect?

It's not a UIQ generation issue, it's a code style issue. They always use quoted includes instead of proper system library includes (as expected by Apple's guidelines) even in the main Qt frameworks themselves.

RytoEX commented:
> > > > > Cannot reproduce - how do you end up in this state?
> > > > 
> > > > 
> > > > @PatTheMav These seem to happen when you have includes like `#include <QtWidgets/QDockWidget>` instead of `#include <QDockWidget>` (the latter doesn't throw warnings). Unfortunately this is what Qt UIC uses when generating the `ui_[...].h` headers.
> > > 
> > > 
> > > Ah yes, Qt generates lots of stuff that is actually wrong on macOS, but we also need to suppress those on obs-studio.
> > > I added suppression for both types of warnings as there is little chance those will be fixed upstream anytime soon.
> > 
> > 
> > Is there an open QTBUG at all documenting that their UIC generation is incorrect?
> 
> It's not a UIQ generation issue, it's a code style issue. They always use quoted includes instead of proper system library includes (as expected by Xcode's default settings) even in the main Qt frameworks themselves.

Even so, that would perhaps be good to document upstream as "this is not what Xcode expects to see" so that they can consider addressing it.

PatTheMav commented:
> > > > > > Cannot reproduce - how do you end up in this state?
> > > > > 
> > > > > 
> > > > > @PatTheMav These seem to happen when you have includes like `#include <QtWidgets/QDockWidget>` instead of `#include <QDockWidget>` (the latter doesn't throw warnings). Unfortunately this is what Qt UIC uses when generating the `ui_[...].h` headers.
> > > > 
> > > > 
> > > > Ah yes, Qt generates lots of stuff that is actually wrong on macOS, but we also need to suppress those on obs-studio.
> > > > I added suppression for both types of warnings as there is little chance those will be fixed upstream anytime soon.
> > > 
> > > 
> > > Is there an open QTBUG at all documenting that their UIC generation is incorrect?
> > 
> > 
> > It's not a UIQ generation issue, it's a code style issue. They always use quoted includes instead of proper system library includes (as expected by Xcode's default settings) even in the main Qt frameworks themselves.
> 
> Even so, that would perhaps be good to document upstream as "this is not what Xcode expects to see" so that they can consider addressing it.

I've done some more research on this and it seems that this requirement stems from the needs for module import support (using `@import Module` by consumers vs directly including headers) as the compiler will figure out which headers to actually include and optimise that usage accordingly, and improved support in Swift.

The major issue I saw is that using `@import` or the suggested `#import` directive both only work for ObjC code. Given that Qt macOS projects can be entirely C++ based, this would make the frameworks incompatible to use for such projects.

gxalpha commented on buildspec.json:
```diff
@@ -1,88 +1,67 @@
     },
+    "tools": {
+        "packages": {
+            "versions": "1.2.10",
```
`versions` might be a typo here? I think the script looks for `version`.

gxalpha commented on .github/scripts/utils.zsh/check_packages:
```diff
@@ -12,41 +12,50 @@ if (( ! ${+commands[packagesbuild]} )) {
+  local version
+  local base_url
+  local hash
+  IFS=';' read -r version base_url hash <<< "$(jq -r '.tools.packages \
+    | {version, baseUrl, hash} \
+    | join(";")' buildspec.json)"
```
I think this fails for me?
macOS CI reports
```
 jq: error: syntax error, unexpected INVALID_CHARACTER, expecting $end (Unix shell quoting issues?) at <top-level>, line 1:
  .tools.packages \                
  jq: 1 compile error
  curl: (3) URL using bad/illegal format or missing URL
  Error: Process completed with exit code 2.
```

gxalpha commented on src/plugin-support.h:
```diff
@@ -16,12 +16,15 @@ You should have received a copy of the GNU General Public License along
+#include <stdlib.h>
+#include <stdarg.h>
+#include <string.h>
 
```
This would benefit from an
```
#ifdef __cplusplus
extern "C" {
#endif
```
for use in C++-plugins. (it's at least needed to make Windows CI happy, macOS was just fine without it for some reason)

PatTheMav commented:
Adressed last set of comments by @gxalpha, removed some architectural redundancy (Qt version and macOS deployment target were specified in build spec and CMake presets file), also updated README file.

RytoEX commented on README.md:
```diff
@@ -2,58 +2,162 @@
+* Encode the .p12 file into its base64 representation by running `base64 <NAME_OF_YOUR_P12_FILE>`
+* Add your developer certificates' names (see above):
+    * `MACOS_SIGNING_APPLICATION_IDENTITY`: Name of the "Developer ID Application" signing certificate
+    *  `MACOS_SIGNING_INSTALLER_IDENTITY`: Name of "Developer ID Installer" signing certificate
```
```suggestion
    * `MACOS_SIGNING_INSTALLER_IDENTITY`: Name of "Developer ID Installer" signing certificate
```

RytoEX commented on README.md:
```diff
@@ -2,58 +2,162 @@
+* Add your developer certificates' names (see above):
+    * `MACOS_SIGNING_APPLICATION_IDENTITY`: Name of the "Developer ID Application" signing certificate
+    *  `MACOS_SIGNING_INSTALLER_IDENTITY`: Name of "Developer ID Installer" signing certificate
+* `MACOS_SIGNING_CERT`: The base64 encoded `.p12` file
+* `MACOS_SIGNING_CERT_PASSWORD`: Password used to generate the .p12 certificate
```
These seem to require additional information about adding them as secrets. They seem a bit detached otherwise.

PatTheMav replied:
The block is explicitly about necessary repository secrets and the first two bullets explain how to generate the file (which requires choosing a password). I don't see what other information might be necessary.

RytoEX replied:
![image](https://github.com/obsproject/obs-plugintemplate/assets/624931/9d9ac352-2e04-4eae-80e2-38a1c80c775a)

The bulleted list items are just a bit disjointed to me. The list is prefaced with:
> To use code signing on GitHub Actions, the certificate and associated information need to be set up as repository secrets in the GitHub repository's settings:

So in that context, these two items make sense. However, they are separated from that preface by intermediate instructions, so I did not immediately connect these two lines to the preface, but instead connected them to the sub-list immediately above.

RytoEX replied:
Looks much better. Thanks!

PatTheMav commented on README.md:
```diff
@@ -2,58 +2,162 @@
+* `author`: Actual name or nickname of the plugin's author
+* `website`: URL of a website associated with the plugin
+* `email`: Contact email address associated with the plugin
+* `uuids`
+    * `macosPackage`: Unique (**!**) identifier for the macOS plugin plugin
+    * `macosInstaller`: Unique (**!**) identifier for the macOS plugin installer
+    * `windowsApp`: Unique (**!**) identifier for the Windows plugin installer
```
I can add this, though I would like to think that "generating a valid UUID" is among the most basic skills somebody should already have when creating a plugin.

RytoEX replied:
I confess, I've never had to think about it. On *nix systems, I suppose one could easily use uuidgen, but I'm not sure what is used on Windows since I've never had to worry about it. Are these UUIDs expected to be v1, v4, or v5? Or does it not matter? Do we want to recommend a specific version?

PatTheMav replied:
The default is random-based UUIDs (v4) to achieve a good-enough level of uniqueness. As long as we don't have a central plugin repository where plugins have to register there is also no way to check for collisions, but UUIDs as random identifiers are common enough (and easy enough to generate via `uuidgen` on Unix systems and `New-Guid` on Windows) for this use case.

(In theory the "bundle id" is actually better, but from what I could tell packages as well as ISS expect UUIDs in their files).

PatTheMav commented on README.md:
```diff
@@ -2,58 +2,162 @@
+    * `qt6`: Prebuilt version of Qt6 as used by OBS Studio
+* `tools`: Contains additional build tools used by CI
+
+The values should be kept in sync with OBS Studio releases and the `buildspec.json` file in use by the main project to ensure that the plugin is developed and built in sync with its target environment.
```
I don't think this requires specific mention - API compatibility is a concern of `obs-studio` and breaking changes would need to be mentioned upon updates there.

Also the plugintemplate is only supposed to be a starting point for plugin development, there is not (and should not) be the idea that plugin developers should rebase their code on this template. 

So it's entirely up to plugin developers to change/adapt this value and its _their_ task to figure out API compatibility and target their plugin accordingly.

RytoEX replied:
Again, I won't block on the mention (or lack thereof) of the caveat in documentation at this time. I just felt it warranted mentioning here. Plugins compiled for older OBS versions _generally_ work fine in newer OBS versions, but plugins compiled for the latest OBS will not load in older versions of OBS if the libobs structs have changed between those versions. It was mostly a point that compiling for the latest version of OBS will lock out users of older OBS versions. While _we_ do not support older versions of OBS, we won't mandate that plugin devs do the same.

> there is not (and should not) be the idea that plugin developers should rebase their code on this template.

Does that mean we should close and reject any Issue submissions that concern older revisions of the plugin template? While it was clear to me that we would not force plugin developers to update/rebase their repositories to this revision, I thought it would be encouraged, as hypothetically most of the files changed are ones they should not have had to modify.

> So it's entirely up to plugin developers to change/adapt this value and its their task to figure out API compatibility and target their plugin accordingly.

Well, yes, that much I agree with. I just wanted something, in the future perhaps, to clarify the ramifications.

PatTheMav replied:
> Again, I won't block on the mention (or lack thereof) of the caveat in documentation at this time. I just felt it warranted mentioning here. Plugins compiled for older OBS versions _generally_ work fine in newer OBS versions, but plugins compiled for the latest OBS will not load in older versions of OBS if the libobs structs have changed between those versions. It was mostly a point that compiling for the latest version of OBS will lock out users of older OBS versions. While _we_ do not support older versions of OBS, we won't mandate that plugin devs do the same.

Got it - adding a short paragraph about "Which Version of OBS Studio To Build With" should address this.

> > there is not (and should not) be the idea that plugin developers should rebase their code on this template.
> 
> Does that mean we should close and reject any Issue submissions that concern older revisions of the plugin template? While it was clear to me that we would not force plugin developers to update/rebase their repositories to this revision, I thought it would be encouraged, as hypothetically most of the files changed are ones they should not have had to modify.

It was my expectation that the template just serves as that: A template. A plugin developer forks it, but then adjusts any and all parts as needed, removing unnecessary code (i.e., Qt6 selection if no UI code is created), removing unnecessary CI stuff (i.e., macOS and Linux build jobs if the plugin only targets Windows systems) and making it "their own".

The problem with just rebasing their local forks is that we can never guarantee that template changes will work with the codebase of an actual in-development plugin. Authors are free to update their own forks that way, but that obviously requires them to know what each part does and not just blindly merge changes.

But what we can't guarantee (and we shouldn't suggest) is to say "just update the C/C++ code, leave everything else as-is and update from upstream". I'd rather say "fork the template, familiarise yourself with all parts of it, customise it to your needs". 

Ideally they should never have to take another look at the template again after - even when it comes how to build `libobs` or pull in dependencies, plugin authors are free to change all of it (because either way, they are responsible for making their customisations work).

RytoEX replied:
Discussed off-thread.

RytoEX commented on README.md:
```diff
@@ -2,58 +2,171 @@
+* `website`: URL of a website associated with the plugin
+* `email`: Contact email address associated with the plugin
+* `uuids`
+    * `macosPackage`: Unique (**!**) identifier for the macOS plugin plugin
```
```suggestion
    * `macosPackage`: Unique (**!**) identifier for the macOS plugin package
```

RytoEX commented on README.md:
```diff
@@ -2,58 +2,171 @@
-The build scripts (contained in the `.github/scripts` directory) will update the `project` line automatically based on values from the `buildspec.json` file. If the scripts are not used, these changes need to be done manually.
+#### Choose An OBS Studio Version
+
+By default the plugin template specifies the most current official OBS Studio version in the `buildspec.json` file, which makes most sense for plugins at the start of development. As far as updating the targeted OBS Studio version is converned, a few things need to be considered:
```
```suggestion
By default the plugin template specifies the most current official OBS Studio version in the `buildspec.json` file, which makes most sense for plugins at the start of development. As far as updating the targeted OBS Studio version is concerned, a few things need to be considered:
```

RytoEX commented on README.md:
```diff
@@ -2,58 +2,171 @@
+
+By default the plugin template specifies the most current official OBS Studio version in the `buildspec.json` file, which makes most sense for plugins at the start of development. As far as updating the targeted OBS Studio version is converned, a few things need to be considered:
+
+* Plugins targeting _older_ version of OBS Studio should _generally_ also work in newer versions, with the exception of breaking changes to specific APIs which would also be explicitly called out in release notes
```
```suggestion
* Plugins targeting _older_ versions of OBS Studio should _generally_ also work in newer versions, with the exception of breaking changes to specific APIs which would also be explicitly called out in release notes
```

RytoEX commented on README.md:
```diff
@@ -2,58 +2,171 @@
+
+* Plugins targeting _older_ version of OBS Studio should _generally_ also work in newer versions, with the exception of breaking changes to specific APIs which would also be explicitly called out in release notes
+* Plugins targeting the _latest_ version of OBS Studio might not work in older versions because the internal data structures used by `libobs` might not be compatible
+* Users are encouraged to always update to the most recent version of OBS Studio available within a reasonable time after release - plugin authors have to chose for themselves if they'd rather keep up with OBS Studio releases or stay with an older version as their baseline (which might of course preclude the plugin from using functionality introduced in a newer version)
```
```suggestion
* Users are encouraged to always update to the most recent version of OBS Studio available within a reasonable time after release - plugin authors have to choose for themselves if they'd rather keep up with OBS Studio releases or stay with an older version as their baseline (which might of course preclude the plugin from using functionality introduced in a newer version)
```

RytoEX commented on README.md:
```diff
@@ -2,58 +2,171 @@
+* Plugins targeting the _latest_ version of OBS Studio might not work in older versions because the internal data structures used by `libobs` might not be compatible
+* Users are encouraged to always update to the most recent version of OBS Studio available within a reasonable time after release - plugin authors have to chose for themselves if they'd rather keep up with OBS Studio releases or stay with an older version as their baseline (which might of course preclude the plugin from using functionality introduced in a newer version)
+
+On Linux the version used for development might be decided by the specific version available via a distribution's package management system, so OBS Studio compatibility for plugins might be determined by those versions instead.
```
```suggestion
On Linux, the version used for development might be decided by the specific version available via a distribution's package management system, so OBS Studio compatibility for plugins might be determined by those versions instead.
```

RytoEX reviewed (changes requested):
Nits

RytoEX commented on .github/actions/build-plugin/action.yaml:
```diff
@@ -0,0 +1,106 @@
+name: 'Set up and build plugin'
+description: 'Builds the plugin for specified architecture and build config.'
```
Nit: this seems to be the only action that uses ending punctuation.

RytoEX reviewed (changes requested):
Nit

gxalpha commented:
Windows and macOS work fine now, Ubuntu appears to fail on CI with Qt enabled:
```
  CMake Warning at cmake/linux/defaults.cmake:83 (_find_package):
    Found package configuration file:
  
      /usr/lib/x86_64-linux-gnu/cmake/Qt6Gui/Qt6GuiConfig.cmake
  
    but it set Qt6Gui_FOUND to FALSE so package "Qt6Gui" is considered to be
    NOT FOUND.  Reason given by package:
  
    Qt6Gui could not be found because dependency WrapOpenGL could not be found.
  
  Call Stack (most recent call first):
    /usr/local/share/cmake-3.26/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
    /usr/lib/x86_64-linux-gnu/cmake/Qt6/QtPublicDependencyHelpers.cmake:14 (find_dependency)
    /usr/lib/x86_64-linux-gnu/cmake/Qt6Widgets/Qt6WidgetsDependencies.cmake:96 (_qt_internal_find_dependencies)
    /usr/lib/x86_64-linux-gnu/cmake/Qt6Widgets/Qt6WidgetsConfig.cmake:50 (include)
    cmake/linux/defaults.cmake:83 (_find_package)
    /usr/lib/x86_64-linux-gnu/cmake/Qt6/Qt6Config.cmake:219 (find_package)
    cmake/linux/defaults.cmake:83 (_find_package)
    cmake/common/helpers_common.cmake:74 (find_package)
    CMakeLists.txt:25 (find_qt)
  
  
  CMake Warning at cmake/linux/defaults.cmake:83 (_find_package):
    Found package configuration file:
  
      /usr/lib/x86_64-linux-gnu/cmake/Qt6Widgets/Qt6WidgetsConfig.cmake
  
    but it set Qt6Widgets_FOUND to FALSE so package "Qt6Widgets" is considered
    to be NOT FOUND.  Reason given by package:
  
    Qt6Widgets could not be found because dependency Qt6Gui could not be found.
  
  Call Stack (most recent call first):
    /usr/lib/x86_64-linux-gnu/cmake/Qt6/Qt6Config.cmake:219 (find_package)
    cmake/linux/defaults.cmake:83 (_find_package)
    cmake/common/helpers_common.cmake:74 (find_package)
    CMakeLists.txt:25 (find_qt)
  
  
  CMake Error at cmake/linux/defaults.cmake:83 (_find_package):
    Found package configuration file:
  
      /usr/lib/x86_64-linux-gnu/cmake/Qt6/Qt6Config.cmake
  
    but it set Qt6_FOUND to FALSE so package "Qt6" is considered to be NOT
    FOUND.  Reason given by package:
  
    Failed to find Qt component "Widgets".
  
    Expected Config file at
    "/usr/lib/x86_64-linux-gnu/cmake/Qt6Widgets/Qt6WidgetsConfig.cmake" exists
  
  
  
  Call Stack (most recent call first):
    cmake/common/helpers_common.cmake:74 (find_package)
    CMakeLists.txt:25 (find_qt)
```

PatTheMav commented:
> Windows and macOS work fine now, Ubuntu appears to fail on CI with Qt enabled:
> 
> ```
>   CMake Warning at cmake/linux/defaults.cmake:83 (_find_package):
>     Found package configuration file:
>   
>       /usr/lib/x86_64-linux-gnu/cmake/Qt6Gui/Qt6GuiConfig.cmake
>   
>     but it set Qt6Gui_FOUND to FALSE so package "Qt6Gui" is considered to be
>     NOT FOUND.  Reason given by package:
>   
>     Qt6Gui could not be found because dependency WrapOpenGL could not be found.
>   
>   Call Stack (most recent call first):
>     /usr/local/share/cmake-3.26/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
>     /usr/lib/x86_64-linux-gnu/cmake/Qt6/QtPublicDependencyHelpers.cmake:14 (find_dependency)
>     /usr/lib/x86_64-linux-gnu/cmake/Qt6Widgets/Qt6WidgetsDependencies.cmake:96 (_qt_internal_find_dependencies)
>     /usr/lib/x86_64-linux-gnu/cmake/Qt6Widgets/Qt6WidgetsConfig.cmake:50 (include)
>     cmake/linux/defaults.cmake:83 (_find_package)
>     /usr/lib/x86_64-linux-gnu/cmake/Qt6/Qt6Config.cmake:219 (find_package)
>     cmake/linux/defaults.cmake:83 (_find_package)
>     cmake/common/helpers_common.cmake:74 (find_package)
>     CMakeLists.txt:25 (find_qt)
>   
>   
>   CMake Warning at cmake/linux/defaults.cmake:83 (_find_package):
>     Found package configuration file:
>   
>       /usr/lib/x86_64-linux-gnu/cmake/Qt6Widgets/Qt6WidgetsConfig.cmake
>   
>     but it set Qt6Widgets_FOUND to FALSE so package "Qt6Widgets" is considered
>     to be NOT FOUND.  Reason given by package:
>   
>     Qt6Widgets could not be found because dependency Qt6Gui could not be found.
>   
>   Call Stack (most recent call first):
>     /usr/lib/x86_64-linux-gnu/cmake/Qt6/Qt6Config.cmake:219 (find_package)
>     cmake/linux/defaults.cmake:83 (_find_package)
>     cmake/common/helpers_common.cmake:74 (find_package)
>     CMakeLists.txt:25 (find_qt)
>   
>   
>   CMake Error at cmake/linux/defaults.cmake:83 (_find_package):
>     Found package configuration file:
>   
>       /usr/lib/x86_64-linux-gnu/cmake/Qt6/Qt6Config.cmake
>   
>     but it set Qt6_FOUND to FALSE so package "Qt6" is considered to be NOT
>     FOUND.  Reason given by package:
>   
>     Failed to find Qt component "Widgets".
>   
>     Expected Config file at
>     "/usr/lib/x86_64-linux-gnu/cmake/Qt6Widgets/Qt6WidgetsConfig.cmake" exists
>   
>   
>   
>   Call Stack (most recent call first):
>     cmake/common/helpers_common.cmake:74 (find_package)
>     CMakeLists.txt:25 (find_qt)
> ```

Looks like Qt requires OpenGL to be installed (which is not correctly reproduced by the package it seems) - installing `libgles2-mesa-dev` seems to have fixed the issue.

So here's the $64k question:

Historically we disabled the frontend-api and Qt by default to not insinuate that plugins are required to be built with either. But as far as CI and this template is concerned, we should enable them both by default to check if the template itself builds correctly.

I'd argue that now that both can be disabled as a CMake option (without code changes necessary), we should enable them by default in the template. Too many issues could've been found if we had compiled it that way on CI by default.

RytoEX commented:
> Historically we disabled the frontend-api and Qt by default to not insinuate that plugins are required to be built with either. But as far as CI and this template is concerned, we should enable them both by default to check if the template itself builds correctly.

Originally, they were _enabled_ by default, which led to some plugins leaving them enabled when they did not use any Qt features, which caused those plugins to break when we updated from Qt 5 to Qt 6 due to linking to Qt. See https://github.com/obsproject/obs-plugintemplate/blob/cdc6df7c3ac35d09c975c5d71db1a44a887fe4dd/CMakeLists.txt#L25-L31.

> I'd argue that now that both can be disabled as a CMake option (without code changes necessary), we should enable them by default in the template. Too many issues could've been found if we had compiled it that way on CI by default.

I'd suggest leaving them disabled by default, but running two separate jobs on CI for our own testing. Mark the second job clearly as "test-frontend-api" or some such, and provide a note inside it that it is not meant to be kept by plugin repos.

Alternatively, we fork the template ourselves into a test repo and build the test jobs there to avoid polluting the template.

CodeYan01 commented:
is 29.1.2 a hard minimum requirement on new plugins going forward? and existing plugins would need to implement this PR on their repo?

PatTheMav commented:
> > Historically we disabled the frontend-api and Qt by default to not insinuate that plugins are required to be built with either. But as far as CI and this template is concerned, we should enable them both by default to check if the template itself builds correctly.
> 
> Originally, they were _enabled_ by default, which led to some plugins leaving them enabled when they did not use any Qt features, which caused those plugins to break when we updated from Qt 5 to Qt 6 due to linking to Qt. See:
> 
> https://github.com/obsproject/obs-plugintemplate/blob/cdc6df7c3ac35d09c975c5d71db1a44a887fe4dd/CMakeLists.txt#L25-L31
> 
> 
> https://github.com/obsproject/obs-plugintemplate/blob/cdc6df7c3ac35d09c975c5d71db1a44a887fe4dd/CMakeLists.txt#L47-L52
> 
> > I'd argue that now that both can be disabled as a CMake option (without code changes necessary), we should enable them by default in the template. Too many issues could've been found if we had compiled it that way on CI by default.
> 
> I'd suggest leaving them disabled by default, but running two separate jobs on CI for our own testing. Mark the test job clearly as "test-frontend-api" or "test-qt" or some such, and provide a note inside it that it is not meant to be kept by plugin repos.
> 
> Alternatively, we fork the template ourselves into a test repo and build the test jobs there to avoid polluting the template.

That would require changing the preset file on the fly in CI jobs as there is no external flag or build script switch to enable this - that would've spread the concern into places where it doesn't belong.

Marking these additional jobs for removal after forking would also make merging upstream changes harder (and goes against the idea that the scripts and CI workflows can be synced from the origin without much care). 

Just changing the default in the preset is simpler, more obvious, and is also precisely the place where this choice should be made and the preset file is literally the first thing people should review/customise.

PatTheMav commented:
> is 29.1.2 a hard minimum requirement on new plugins going forward? and existing plugins would need to implement this PR on their repo?

New plugins should always target the current stable version, but authors are free to change the version and source hashes as they need with the caveats mentioned in the updated readme file.

CodeYan01 commented:
thanks

CodeYan01 commented:
I suggest adding steps in the readme for building for local development. Currently, only github actions are explained, but plugin authors would want to build it for themselves locally.

RytoEX commented:
> Just changing the default in the preset is simpler, more obvious, and is also precisely the place where this choice should be made and the preset file is literally the first thing people should review/customise.

Unfortunately, evidence shows that they don't.

gxalpha reviewed (approved):
This works on my machine for my plugin, as well as on CI at least to the point where I use "new" OBS APIs and Ubuntu fails. But Ubuntu's packages are ancient so nothing doing about that I suppose.

RytoEX reviewed (approved):
Looks good to me.

paulpv commented on cmake/windows/resources/installer-Windows.iss.in:
```diff
@@ -7,7 +7,7 @@
 ; Do not use the same AppId value in installers for other applications.
 ; (To generate a new GUID, click Tools | Generate GUID inside the IDE.)
-AppId={{CD703FE5-1F2C-4837-BD3D-DD840D83C3E3}
+AppId={{@UUID_APP@}
```
@PatTheMav is either the original or updated (and still current) `{{` a typo, or is the double open curly brace correct?
